test/categorical_algebra/DataMigration.jl

```diff
@@ -0,0 +1,180 @@
+using Catlab.Present
+
+# Pullback data migration
+###########################
+
+@present TheoryDDS(FreeSchema) begin
+  X::Ob
+  Φ::Hom(X,X)
+end
+
+const AbstractDDS = AbstractCSetType(TheoryDDS)
+const DDS = CSetType(TheoryDDS, index=[:Φ])
+
+h = Graph(3)
+add_parts!(h, :E, 3, src = [1,2,3], tgt = [2,3,1])
+
+# Identity data migration.
+@test h == Graph(h, Dict(:V => :V, :E => :E),
+                    Dict(:src => :src, :tgt => :tgt))
+
+# Migrate DDS → Graph.
+dds = DDS()
+add_parts!(dds, :X, 3, Φ=[2,3,1])
+X = TheoryDDS[:X]
+@test h == Graph(dds, Dict(:V => :X, :E => :X),
+                 Dict(:src => id(X), :tgt => :Φ))
+
+h2 = copy(h)
+migrate!(h2, dds, Dict(:V => :X, :E => :X),
+                  Dict(:src => id(X), :tgt => :Φ))
+@test h2 == ob(coproduct(h, h))
+
+# Migrate DDS → DDS by advancing four steps.
+@test dds == DDS(dds, Dict(:X => :X),
+                      Dict(:Φ => [:Φ, :Φ, :Φ, :Φ]))
+
+@present TheoryLabeledDDS <: TheoryDDS begin
+  Label::Data
+  label::Attr(X, Label)
+end
+const LabeledDDS = ACSetType(TheoryLabeledDDS, index=[:Φ, :label])
+
+S, ϕ, Label, label = generators(TheoryLabeledDDS)
+V, E, s, t, Weight, weight = generators(TheoryWeightedGraph)
+
+ldds = LabeledDDS{Int}()
+add_parts!(ldds, :X, 4, Φ=[2,3,4,1], label=[100, 101, 102, 103])
+
+wg = WeightedGraph{Int}(4)
+add_parts!(wg, :E, 4, src=[1,2,3,4], tgt=[2,3,4,1], weight=[101, 102, 103, 100])
+
+@test wg == WeightedGraph{Int}(ldds,
+  Dict(:V => :X, :E => :X),
+  Dict(:src => id(S), :tgt => :Φ, :weight => [:Φ, :label]))
+
+@test Presentation(Graph(1)) == TheoryGraph
+@test Presentation(ldds) == TheoryLabeledDDS
+
+
+F = Functor(
+  Dict(V => S, E => S), 
+  Dict(s => id(S), t => ϕ, weight => compose(ϕ, label)),
+  TheoryWeightedGraph, TheoryLabeledDDS
+)
+
+@test wg == WeightedGraph{Int}(ldds, F) 
+
+
+# Left Pushforward data migration
+
+#################################
+
+Σ = Sigma
+
+@present ThBipartite(FreeSchema) begin
+  (V1, V2, E)::Ob
+  src::Hom(E, V1)
+  tgt::Hom(E, V2)
+end
+
+V1B, V2B, EB = generators(ThBipartite, :Ob)
+srcB, tgtB = generators(ThBipartite, :Hom)
+
+VG, EG = generators(TheoryGraph, :Ob)
+srcG, tgtG = generators(TheoryGraph, :Hom)
+
+F = Functor(
+    Dict(V1B => VG, V2B => VG, EB => EG), 
+    Dict(srcB => srcG, tgtB => tgtG),
+    ThBipartite, TheoryGraph
+)
+
+idF = Functor(
+  Dict(VG => VG, EG => EG),
+  Dict(srcG => srcG, tgtG => tgtG),
+  TheoryGraph, TheoryGraph
+)
+
+ΣF =  Σ(F)
+X = ACSetType(ThBipartite, index=[:src, :tgt])()
+Y = ΣF(X)
+
+@test nparts(Y, :V) == 0
+@test nparts(Y, :E) == 0
+
+X =  @acset ACSetType(ThBipartite, index=[:src, :tgt]) begin
+  V1 = 4
+  V2 = 3
+  E = 4
+  src = [1,2,2,3]
+  tgt = [1,1,2,3]
+end
+
+Y = ΣF(X)
+@test nparts(Y, :V) == 7
+@test nparts(Y, :E) == 4
+@test length(Y[:src] ∩ Y[:tgt]) == 0
+
+@test Σ(idF)(Y) == Y
+
+@present ThSpan(FreeSchema) begin 
+  (L1, L2, A)::Ob
+  l1::Hom(A, L1)
+  l2::Hom(A, L2)
+end
+
+X = @acset ACSetType(ThSpan, index=[:l1, :l2]) begin
```

Comment:
Especially since future versions of `ACSetType` may involve more metaprogramming, I think we should write

```julia
const Span = ACSetType(ThSpan, index=[:l1, :l2])
```

and avoid putting `ACSetType` calls inline. (Actually, I'm surprised the above works, but that's cool I guess :)